kratos/utilities/nurbs_utilities/snake_sbm_utilities.cpp

```diff
@@ -0,0 +1,872 @@
+        }
+        DynamicBins testBinInner(points.begin(), points.end());
+
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && is_inner) << "Inner :: Ending SnakeStep" << std::endl;
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && !is_inner) << "Outer :: Ending SnakeStep" << std::endl;
+        
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && is_inner) << "Inner :: Starting MarkKnotSpansAvailable" << std::endl;
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && !is_inner) << "Outer :: Starting MarkKnotSpansAvailable" << std::endl;
+
+        for (IndexType i = 0; i < numberOfLoops; i++) {
+            IndexType idInnerLoop = i;
+            // Mark the knot_spans_available's for inner and outer loops
+            MarkKnotSpansAvailable(knot_spans_available, idInnerLoop, testBinInner, r_skin_sub_model_part, lambda, 
+                                   n_knot_spans_uv, knot_step_uv, starting_pos_uv);  
+            
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && is_inner) << "Inner :: Ending MarkKnotSpansAvailable" << std::endl;
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && !is_inner) << "Outer :: Ending MarkKnotSpansAvailable" << std::endl;
+            
+            if (is_inner) {
+                CreateSurrogateBuondaryFromSnakeInner(knot_spans_available, idInnerLoop, r_surrogate_sub_model_part, 
+                                                      r_skin_sub_model_part, testBinInner, n_knot_spans_uv, 
+                                                      knotVectorU, knotVectorV);
+
+                KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0) << "Inner :: Snake process has finished" << std::endl;
+            }
+            else {
+                CreateSurrogateBuondaryFromSnakeOuter (knot_spans_available, idInnerLoop, r_surrogate_sub_model_part,
+                                                        r_skin_sub_model_part, testBinInner, n_knot_spans_uv, knotVectorU,
+                                                        knotVectorV, starting_pos_uv);
+                                                        
+                KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0) << "Outer :: Snake process has finished" << std::endl;
+            }
+        }
+
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && is_inner) << "Inner :: Loop finished" << std::endl;
+        KRATOS_INFO_IF("::[SnakeSbmUtilities]::", r_echo_level > 0 && !is_inner) << "Outer :: Loop finished" << std::endl;
+    }
+
+
+    void SnakeSbmUtilities::SnakeStep(
+        ModelPart& rSkinModelPart, 
+        std::vector<std::vector<std::vector<int>>> &knotSpansAvailable, 
+        int idMatrix, 
+        std::vector<std::vector<int>> knotSpansUV, 
+        std::vector<std::vector<double>> conditionCoord, 
```

Comment:
Wouldn't it be easier to pass the condition and handle the coordinates inside the function?

Note that this would avoid allocating such matrix.